### PR TITLE
Change terminology used for input output parameters

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description

As discussed here https://github.com/ethereum/solidity/pull/4859#discussion_r214986820 this first PR changes the terminology previously used for input and output parameters. Hopefully, I got the semantics, explanations and usage correct. Part 2 is here https://github.com/ethereum/solidity/pull/5324.

### CODE REVIEW

**Comment by lazaridiscom** [comment]
"Input" and "Output" are still used, which reads sometimes strange. see inline  comments.

File: docs/control-structures.rst [hunk 1]
```diff
@@ -2,23 +2,22 @@
...
-----------------
+Function Parameters
+-------------------
 
-The input parameters are declared the same way as variables are.
-The name of unused parameters can be omitted.
-For example, suppose we want our contract to
-accept one kind of external calls with two integers, we would write
-something like::
+Input function parameters are declared the same way as variables, and the name of
```
[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply docs/control-structures.rst]: Yes, this was somewhat intentional @lazaridiscom mostly to introduce the terminology first. I'm not committed to the idea though. Let's see what @chriseth thinks.

File: docs/control-structures.rst [hunk 2]
```diff
@@ -29,16 +28,16 @@ something like::
...
-Input parameters can be used just as any other local variable
-can be used, they can also be assigned to.
+Input parameters can be used as any other local variable and they can also be assigned to.
+
+Return Types
+------------
 
-Output Parameters
------------------
+Output function return types are declared with the same syntax after the
```
[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply docs/control-structures.rst]: ditto

File: docs/control-structures.rst [hunk 3]
```diff
@@ -29,16 +28,16 @@ something like::
         }
     }
 
-Input parameters can be used just as any other local variable
-can be used, they can also be assigned to.
+Input parameters can be used as any other local variable and they can also be assigned to.
```
[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply docs/control-structures.rst]: @lazaridiscom That one is a catch :)

File: docs/control-structures.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (lazaridiscom) [reply docs/control-structures.rst]: Fine. Just be aware that "Input function parameters" is simply misleading. Same below.

File: docs/control-structures.rst [hunk 4]
```diff
@@ -2,23 +2,22 @@
 Expressions and Control Structures
 ##################################
 
-.. index:: ! parameter, parameter;input, parameter;output, parameter;multiple
+.. index:: ! parameter, parameter;input, parameter;output, function parameters, parameter;function, return types, types;return, multiple
 
-Input Parameters and Output Parameters
-======================================
+Function Parameters and Return Types
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
Function Parameters and Return Variables
```

File: docs/control-structures.rst [hunk 5]
```diff
@@ -2,23 +2,22 @@
...
+====================================
 
-As in Javascript, functions may take parameters as input;
-unlike in Javascript and C, they may also return arbitrary number of
-parameters as output.
+As in JavaScript, functions may take parameters as input. Unlike in JavaScript
+and C, functions may also return an arbitrary number of types as output.
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
and C, functions may also return an arbitrary number of values as output.
```

File: docs/control-structures.rst [hunk 6]
```diff
@@ -2,23 +2,22 @@
 Expressions and Control Structures
 ##################################
 
-.. index:: ! parameter, parameter;input, parameter;output, parameter;multiple
+.. index:: ! parameter, parameter;input, parameter;output, function parameters, parameter;function, return values, types;return, multiple
 
-Input Parameters and Output Parameters
-======================================
+Function Parameters and Return Values
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
Function Parameters and Return Variables
```

File: docs/control-structures.rst [hunk 7]
```diff
@@ -55,12 +54,12 @@ write::
         }
     }
 
-The names of output parameters can be omitted.
-The return values can be specified using ``return`` statements,
-which are also capable of :ref:`returning multiple values<multi-return>`.
+The names of return variables can be omitted.
+The return values can also be specified using ``return`` statements,
+which are capable of :ref:`returning multiple values<multi-return>`.
 Return parameters can be used as any other local variable and they
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
Return variables can be used as any other local variable and they
```

File: docs/control-structures.rst [hunk 8]
```diff
@@ -55,12 +54,31 @@ write::
...
+            public
+            pure
+            returns (uint o_sum, uint o_product)
+        {
+            return (_a + _b, _a * _b);
+        }
+    }
+
+This form is equivalent to first assigning values to the
+return variables and then using ``return;;`` to leave the function.
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
return variables and then using ``return;`` to leave the function.
```

File: docs/control-structures.rst [hunk 9]
```diff
@@ -2,23 +2,22 @@
 Expressions and Control Structures
 ##################################
 
-.. index:: ! parameter, parameter;input, parameter;output, parameter;multiple
+.. index:: ! parameter, parameter;input, parameter;output, function parameters, parameter;function, return values, types;return, multiple
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
.. index:: ! parameter, parameter;input, parameter;output, function parameters, parameter;function, return variable, variable;return, return
```

File: docs/control-structures.rst (hunk 5 above)
Comment (lazaridiscom) [reply docs/control-structures.rst]: > functions may also return an arbitrary number of values as output

Seeing the use of "Return Variables" further below, shouldn't this become something like that:

=> functions may also return an arbitrary number of variables as output. Those return variables can have different types.

**Review by chriseth — APPROVED** [review]

File: docs/control-structures.rst (hunk 5 above)
Comment (chriseth) [reply docs/control-structures.rst]: I thought it would be obvious that they can have different types, but we can add it!

Comment (chriseth) [reply docs/control-structures.rst]: Actually this changed quite a bit, could you check the current version again please?

File: docs/control-structures.rst [hunk 10]
```diff
@@ -2,23 +2,22 @@
...
-Input Parameters and Output Parameters
-======================================
+Function Parameters and Return Variables
+========================================
 
-As in Javascript, functions may take parameters as input;
-unlike in Javascript and C, they may also return arbitrary number of
-parameters as output.
+As in JavaScript, functions may take parameters as input. Unlike in JavaScript
+and C, functions may also return an arbitrary number of values as output.
```
Comment (lazaridiscom) [inline docs/control-structures.rst]: possibly stick to "variables" (instead of "values", as in rest of document)

Comment (chriseth) [reply docs/control-structures.rst]: I would argue that the stuff you return are values, but you do so with the help of return variables.

Comment (lazaridiscom) [reply docs/control-structures.rst]: I don't know how you return stuff. The issue here is "Consistency of Documentation". You introduce "Return Variables", and use the term "Variables" in the rest of the document, so "Values"  feels a bit inconsistent at this point.

Comment (lazaridiscom) [reply docs/control-structures.rst]: Just went through the whole document. It seems the easiest would be to to merge this as-is, and then

- Clarify difference of "Return Value" and "Return Variable" (within the document)
- use only one(!) term throughout the document/documentation (personally I would prefer "Return Value", though me not knowing if its technically the correct term).

Comment (chriseth) [reply docs/control-structures.rst]: To me, "return value" and "return variable" are actually two different things.

**Comment by chriseth** [comment]
We also wanted to move this section to a different place, so I'll merge for now.

File: docs/control-structures.rst (hunk 10 above)
Comment (lazaridiscom) [reply docs/control-structures.rst]: Of course its different, we agree here. I just mentioned the need to clarify this.

follow-up: #5425

**Pull request merged**